src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -747,6 +748,66 @@ private bool CheckFieldValueKind(SyntaxNode node, BoundFieldAccess fieldAccess,
             return CheckIsValidReceiverForVariable(node, fieldAccess.ReceiverOpt, valueKind, diagnostics);
         }
 
+        bool IsAllowedInitOnlyAssignment(Symbol accessedMember, BoundExpression receiver, bool restrictToThis)
+        {
+            // ok: new C() { InitOnlyProperty = ... }
+            if (receiver is BoundObjectOrCollectionValuePlaceholder)
+            {
+                return true;
+            }
+
+            // bad: other.InitOnlyProperty = ...
+            if (!(receiver is BoundThisReference) && !(receiver is BoundBaseReference) && !(receiver is BoundObjectOrCollectionValuePlaceholder))
+            {
+                return false;
+            }
+
+            var containingMember = ContainingMemberOrLambda;
+            if (!(containingMember is MethodSymbol method))
+            {
+                return false;
+            }
+
+            if (!isDerivedType(method.ContainingType, accessedMember.ContainingType.OriginalDefinition, restrictToThis))
```

Comment:
>if (!isDerivedType(method.ContainingType, accessedMember.ContainingType.OriginalDefinition, restrictToThis)) [](start = 12, length = 108)

It is not clear what are we trying to accomplish with this check. Please add comments explaining what are we trying to enforce as a whole and on each step.

Reply:
I think I can now adjust the receiver check above to simplify this. Will do.

---
In reply to: [411009367](https://github.com/dotnet/roslyn/pull/43275#discussion_r411009367) [](ancestors = 411009367)